asteroid/engine/system.py

```diff
@@ -150,7 +158,41 @@ def validation_epoch_end(self, outputs):
         """
         avg_loss = torch.stack([x["val_loss"] for x in outputs]).mean()
         tensorboard_logs = {"val_loss": avg_loss}
-        return {"val_loss": avg_loss, "log": tensorboard_logs, "progress_bar": tensorboard_logs}
+        return {"val_loss": avg_loss, "log": tensorboard_logs,
+                "progress_bar": tensorboard_logs}
+
+    def test_step(self, batch, batch_nb):
+
+        inputs, targets, sizes = batch
+        est_targets = self(inputs)
+        loss, reordered_sources = self.loss_func(est_targets, targets,
+                                                 return_est=True)
+        utt_metrics = batch_metrics(
+            inputs,
+            targets,
+            reordered_sources,
+            sizes,
+            sample_rate=self.sample_rate,
+            metrics_list=self.metric_list,
+        )
+        return utt_metrics
+
+    def test_epoch_end(self, outputs):
```

Comment:
Here it would be really good to know what kind of thing `outputs` is. From reading this method's code only I have no clue. Maybe a combination of some of the following things would be helpful:

- docstring
- using tuple unpacking rather than `[0]`, `[1]` etc, for example `foo, *bars = outputs`
- using names more descriptive than `key` and `x` ;)

Reply:
I fully agree with that. It was not clear at first sight and I had to read the rest of the code to get it.